File: ios/App/Snappet/Features/Journal/JournalEntry.swift
import Foundation
import SwiftData

/// One journal entry in the shared store. The integrator appends this type to
/// `SnappetSchema.models`.
///
/// Note: `body` is a stored `String` here, not the SwiftUI `View.body` — this is a
/// `@Model` class, not a `View`, so the name does not clash.
@Model
final class JournalEntry {
    /// Optional user-provided title. May be empty.
    var title: String
    /// The entry's free-form text.
    var body: String
    /// When the entry was first created.
    var createdAt: Date
    /// When the entry was last edited (equals `createdAt` until first edit).
    var updatedAt: Date
    /// Free-form tags for organizing/filtering. Stored normalized (see `normalizeTags`):
    /// trimmed, lowercased, de-duplicated, empties dropped. Defaulted so SwiftData performs a
    /// lightweight additive migration for stores created before this property existed.
    var tags: [String] = []

    init(title: String, body: String, createdAt: Date = .now, updatedAt: Date = .now, tags: [String] = []) {
        self.title = title
        self.body = body
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.tags = JournalEntry.normalizeTags(tags)
    }

    /// Whether an entry with these fields carries no user content at all — the shared
    /// definition behind both the editor's Done guard and the abandoned-entry cleanup,
    /// so "blank" can't drift between the two paths.
    static func isBlank(title: String, body: String, tags: [String]) -> Bool {
        title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            && body.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            && tags.isEmpty
    }

    /// An abandoned pre-inserted row from `createEntry()`: blank **and** never Done-saved
    /// (`save()` is the only place `updatedAt` moves off `createdAt`). The extra clause keeps
    /// the appear-time sweep from ever touching a real entry the user deliberately emptied.
    static func isAbandonedBlank(_ entry: JournalEntry) -> Bool {
        isBlank(title: entry.title, body: entry.body, tags: entry.tags)
            && entry.updatedAt == entry.createdAt
    }

    /// Normalize raw tag strings: trim whitespace, lowercase, drop empties, and de-duplicate
    /// while preserving first-seen order.
    static func normalizeTags(_ raw: [String]) -> [String] {
        var seen = Set<String>()
        var result: [String] = []
        for tag in raw {
            let normalized = tag.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
            guard !normalized.isEmpty, seen.insert(normalized).inserted else { continue }
            result.append(normalized)
        }
        return result
    }
}

File: android/app/src/main/java/com/snappet/mobile/feature/journal/JournalEntry.kt
package com.snappet.mobile.feature.journal

import androidx.room.Entity
import androidx.room.PrimaryKey

/**
 * One journal entry in the shared store. Mirrors the iOS `JournalEntry` `@Model`.
 *
 * Tags are stored as a single comma-joined [String] (normalized: trimmed, lowercased,
 * de-duplicated, empties dropped) to avoid needing a Room `TypeConverter` for `List`.
 * Use [tagList] / [joinTags] / [normalizeTags] to move between the two representations.
 */
@Entity(tableName = "journal_entries")
data class JournalEntry(
    @PrimaryKey(autoGenerate = true) val id: Long = 0,
    /** Optional user-provided title. May be empty. */
    val title: String,
    /** The entry's free-form text. */
    val body: String,
    /** Comma-joined, normalized tags (lowercased, de-duplicated). May be empty. */
    val tags: String,
    /** When the entry was first created (epoch millis). Drives newest-first ordering. */
    val createdAt: Long,
    /** When the entry was last edited (epoch millis; equals [createdAt] until first edit). */
    val updatedAt: Long,
) {
    /** The stored [tags] string split back into a normalized list. */
    val tagList: List<String>
        get() = if (tags.isBlank()) emptyList() else tags.split(",").filter { it.isNotEmpty() }

    companion object {
        /**
         * Normalize raw tag strings: trim whitespace, lowercase, drop empties, and de-duplicate
         * while preserving first-seen order. Mirrors iOS `JournalEntry.normalizeTags`.
         */
        fun normalizeTags(raw: List<String>): List<String> {
            val seen = HashSet<String>()
            val result = ArrayList<String>()
            for (tag in raw) {
                val normalized = tag.trim().lowercase()
                if (normalized.isEmpty() || !seen.add(normalized)) continue
                result.add(normalized)
            }
            return result
        }

        /** Join a list of tags into the stored comma-separated form (after normalizing). */
        fun joinTags(raw: List<String>): String = normalizeTags(raw).joinToString(",")
    }
}
